### PR TITLE
common/buffer: fix stack corruption in rebuild_aligned_size_and_memory()

### PR DESCRIPTION
**Pull request by CongMinYin**
There is such a bl, which needs to satisfy two conditions:

1. all ptr's length sum except last ptr is aligned with 4K;
2. the length of last ptr is 0.

This bl will cause stack corruption when calling
bufferlist::rebuild_aligned_size_and_memory().
    
Deal with this special scenario in rebuild_aligned_size_and_memory() to
solve the bug. And added a specialtest-case to reproduce this scenario.
    
Fixes: https://tracker.ceph.com/issues/51419

Signed-off-by: Yin Congmin <congmin.yin@intel.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by CongMinYin** [comment]
Hi @rzarzynski , in the process of solving the RBD bug, I found there was a problem with the common buffer. I have written a test case to reproduce this problem. Can you have a look at it.

**Comment by CongMinYin** [comment]
CC @idryomov , new issue about pwl/ssd, this to fix https://tracker.ceph.com/issues/51419.

**Comment by CongMinYin** [comment]
CC @majianpeng

**Comment by tchaikov** [comment]
@CongMinYin could you summarize the change you made in the commit in the title. see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#commit-title 

"fix SIGABRT" is the effect of this change, not it does not describe the change.

**Comment by idryomov** [comment]
Yes, please.  This needs a detailed explanation of what is wrong with the `bufferlist` code.

Separately, I would appreciate if you describe the RBD bug that led you to discovering this.

File: src/test/bufferlist.cc [hunk 1]
```diff
@@ -2369,6 +2369,21 @@ TEST(BufferList, splice) {
...
     EXPECT_EQ(0, ::memcmp("fghi", other.c_str(), other.length()));
   }
+
+  {
+    bufferlist bl1, bl;
+    std::vector<int> data(30);
+    // must use create_aligned
+    bufferptr ptr(buffer::create_aligned(16384, 4096));
+
+    bl.clear();
```
Comment (idryomov) [inline src/test/bufferlist.cc:2379]: Is this needed?  Shouldn't `bl` be clear at this point?

Just making sure I understand the test case.

**Comment by majianpeng** [comment]
use our test case w/o this pr;
```
before splice blbuffer::list(len=20480,
        buffer::ptr(0~124 0x563a79612d00 in raw 0x563a79612d00 len 124 nref 1),
        buffer::ptr(0~3972 0x563a7966c010 in raw 0x563a7966c010 len 8096 nref 1),
        buffer::ptr(0~16384 0x563a79668000 in raw 0x563a79668000 len 16384 nref 2)
)
after splice blbuffer::list(len=12288,
        buffer::ptr(4096~12288 0x563a79669000 in raw 0x563a79668000 len 16384 nref 3),
        buffer::ptr(3972~0 0x563a7966cf94 in raw 0x563a7966c010 len 8096 nref 2)
```
in func rebuild_aligned_size_and_memory, it will corrupt stack. This because
a)bl has a ptr which length is zero.
b)the first ptr's size and memroy-address aligned w/ 4K. 
So  https://github.com/ceph/ceph/blob/master/src/common/buffer.cc#L1261 will corrupt stack and cause SIGABRT.

File: src/common/buffer.cc [hunk 2]
```diff
@@ -1645,7 +1645,6 @@ static ceph::spinlock debug_lock;
         curbuf = _buffers.erase_after(curbuf_prev);
 	_carriage->set_offset(_carriage->offset() + _carriage->length());
 	_carriage->set_length(0);
-	_buffers.push_back(*_carriage);
```
Comment (rzarzynski) [inline src/common/buffer.cc:1648]: Well, if we really need to drop this line, the comment above should be reworked as well. Also, I'm worried the patch introduces a memory leak as `erase_after()` doesn't dispose the node.

In the meantime, I joining the folks asking for more descriptive commit message.

File: src/test/bufferlist.cc (hunk 1 above)
Comment (rzarzynski) [reply src/test/bufferlist.cc:2379]: Looks like no-operation to me.

**Review by rzarzynski — CHANGES_REQUESTED** [review]

File: src/test/bufferlist.cc (hunk 1 above)
Comment (CongMinYin) [reply src/test/bufferlist.cc:2379]: > Is this needed? Shouldn't `bl` be clear at this point?
> 
> Just making sure I understand the test case.

Not needed, I will delete this line later.

File: src/common/buffer.cc (hunk 2 above)
Comment (CongMinYin) [reply src/common/buffer.cc:1648]: > Well, if we really need to drop this line, the comment above should be reworked as well. Also, I'm worried the patch introduces a memory leak as `erase_after()` doesn't dispose the node.
> 
> In the meantime, I joining the folks asking for more descriptive commit message.

I'm not sure about the solution, but the problem does exist. see the issue describe by Jianpeng

Comment (rzarzynski) [reply src/common/buffer.cc:1648]: @CongMinYin, @majianpeng: I'm not questioning that at all. :-) Just arguing about the way how we fix it.

>  a)bl has a ptr which length is zero.

This is still valid and can (rarely) happen because of how `_carriage` and the non-pointer `append()` variants of `contiguous_appender` interact:

```cpp
    class contiguous_appender {
      // ...
      void append(const bufferlist& l) {
        if (deep) {
          // ...
        } else {
          flush_and_continue();
          bl.append(l);
          space = bl.obtain_contiguous_space(0);
          out_of_band_offset += l.length();
        }
      }
```

My impression is we need to take a look on `rebuild_aligned_size_and_memory()`. I can help with that.

Comment (rzarzynski) [reply src/common/buffer.cc:1648]: Also, take a look on the `BufferListIterator,iterate_with_empties` testcase.

Comment (CongMinYin) [reply src/common/buffer.cc:1648]: We also look forward to discussing with you and seeking the most appropriate solution.

Comment (rzarzynski) [reply src/common/buffer.cc:1648]: OK, I think the fix could look like this one:

```diff
diff --git a/src/common/buffer.cc b/src/common/buffer.cc
index 548abc458fd..9c76cde6eed 100644
--- a/src/common/buffer.cc
+++ b/src/common/buffer.cc
@@ -1258,8 +1258,12 @@ static ceph::spinlock debug_lock;
             buffer::create_aligned(unaligned._len, align_memory)));
         had_to_rebuild = true;
       }
-      _buffers.insert_after(p_prev, *ptr_node::create(unaligned._buffers.front()).release());
-      _num += 1;
+      if (unaligned.get_num_buffers()) {
+        _buffers.insert_after(p_prev, *ptr_node::create(unaligned._buffers.front()).release());
+        _num += 1;
+      } else {
+        // a bufferlist containing only 0-length bptrs is rebuild as empty
+      }
       ++p_prev;
     }
     return had_to_rebuild;
```

**Comment by MahatiC** [comment]
> Separately, I would appreciate if you describe the RBD bug that led you to discovering this.

@idryomov an example where this can be reproduced is when the below fio config is run

```
[global]

ioengine=rbd
clientname=admin
rw=randwrite
#bs=1m
bs=16k
time_based=1
runtime=20s
iodepth=16
pool=rbd
group_reporting

[volumes]
rbdname=fio-test
```

File: src/common/buffer.cc (hunk 2 above)
Comment (CongMinYin) [reply src/common/buffer.cc]: Thank you for your solution. I have updated the code and commit information of PR.

File: src/test/bufferlist.cc [hunk 3]
```diff
@@ -2369,6 +2369,20 @@ TEST(BufferList, splice) {
     EXPECT_EQ((unsigned)11, bl.length());
     EXPECT_EQ(0, ::memcmp("fghi", other.c_str(), other.length()));
   }
+
+  {
+    bufferlist bl1, bl;
```
Comment (rzarzynski) [inline src/test/bufferlist.cc:2374]: It would be helpful to have a comment here explaining the block is actually a bug replicator. Also, there is no any `EXPECT_...` clause.

Comment (CongMinYin) [reply src/test/bufferlist.cc]: > It would be helpful to have a comment here explaining the block is actually a bug replicator. Also, there is no any `EXPECT_...` clause.

done, please check again.

**Comment by idryomov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-07-01_14:09:56-rados-wip-kefu-testing-2021-07-01-1912-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-07-02_03:53:50-rados-wip-kefu-testing-2021-07-02-0958-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-07-02_12:31:26-rados-wip-kefu-testing-2021-07-02-1417-distro-basic-smithi/

failures are not related to this changeset.

**Comment by tchaikov** [comment]
@rzarzynski could you take another look?

File: src/test/bufferlist.cc [hunk 4]
```diff
@@ -2369,6 +2369,22 @@ TEST(BufferList, splice) {
...
+    bl.clear();
+    std::vector<int> data(30);
+    bufferptr ptr(buffer::create_aligned(16384, 4096));
+
+    encode(data, bl);
+    bl.append_zero(4096 - bl.length());
+    bl.append(ptr);
+    bl.splice(0, 8192, &bl1);
+    bl.rebuild_aligned_size_and_memory(4096, 4096, 1024);
+  }
```
Comment (rzarzynski) [inline src/test/bufferlist.cc:2386]: Oops, looks an `EXPECT_...` clause is still missed. Not staged when amending the commit?

**Review by rzarzynski — APPROVED** [review]

File: src/test/bufferlist.cc (hunk 4 above)
Comment (CongMinYin) [reply src/test/bufferlist.cc:2389]: > Oops, looks an `EXPECT_...` clause is still missed. Not staged when amending the commit?

sorry, not familiar with this process. Now `EXPECT_...` has been added. Please check is it ok?

Comment (idryomov) [reply src/test/bufferlist.cc:2389]: I think this should be moved to `TEST(BufferList, rebuild_aligned_size_and_memory)`.  It doesn't belong in `TEST(BufferList, splice)` because `splice()` is not the problem / what is being tested here.

File: src/test/bufferlist.cc [hunk 5]
```diff
@@ -2369,6 +2369,25 @@ TEST(BufferList, splice) {
...
+    std::vector<int> data(30);
+    bufferptr ptr(buffer::create_aligned(16384, 4096));
+
+    encode(data, bl);
+    bl.append_zero(4096 - bl.length());
+    bl.append(ptr);
+    bl.splice(0, 8192, &bl1);
+    EXPECT_EQ(bl.get_num_buffers(), 2);
+    EXPECT_EQ(bl.back().length(), 0);
+    bl.rebuild_aligned_size_and_memory(4096, 4096, 1024);
```
Comment (idryomov) [inline src/test/bufferlist.cc:2387]: Is `max_buffers=1024` relevant for reproducing?  Wouldn't `bl.rebuild_aligned(4096)` be sufficient?

File: src/test/bufferlist.cc (hunk 4 above)
Comment (idryomov) [reply src/test/bufferlist.cc:2389]: I also wonder if it might be better to avoid `splice()` in the reproducer and construct the bufferlist matching the conditions (two buffers, first is memory and size aligned, second is 0-length) manually?  That way this test case would still work even if the implementation of `splice()` changes in the future.

**Comment by CongMinYin** [comment]
@idryomov made changes as you suggested, please check.

File: src/common/buffer.cc [hunk 6]
```diff
@@ -1258,8 +1258,12 @@ static ceph::spinlock debug_lock;
             buffer::create_aligned(unaligned._len, align_memory)));
         had_to_rebuild = true;
       }
-      _buffers.insert_after(p_prev, *ptr_node::create(unaligned._buffers.front()).release());
-      _num += 1;
+      if (unaligned.get_num_buffers()) {
+        _buffers.insert_after(p_prev, *ptr_node::create(unaligned._buffers.front()).release());
+        _num += 1;
+      } else {
+        // a bufferlist containing only 0-length bptrs is rebuild as empty
```
Comment (idryomov) [inline src/common/buffer.cc:1265]: Nit: rebuilt

File: src/test/bufferlist.cc [hunk 7]
```diff
@@ -1848,6 +1848,22 @@ TEST(BufferList, rebuild_aligned_size_and_memory) {
   EXPECT_TRUE(bl.is_aligned(SIMD_ALIGN));
   EXPECT_TRUE(bl.is_n_align_sized(BUFFER_SIZE));
   EXPECT_EQ(3U, bl.get_num_buffers());
+
+  {
+    /* bug replicator, to test rebuild_aligned_size_and_memory() in this
+     * special scenarios */
```
Comment (idryomov) [inline src/test/bufferlist.cc:1854]: Nit: ... in the scenario where the first bptr is both size and memory aligned and the second is 0-length.

File: src/test/bufferlist.cc [hunk 8]
```diff
@@ -1848,6 +1848,22 @@ TEST(BufferList, rebuild_aligned_size_and_memory) {
...
+    bl.clear();
+    bufferptr ptr1(buffer::create_aligned(4096, 4096));
+    bl.append(ptr1);
+    bufferptr ptr(10);
+    /* bl.back().length() must be 0 */
+    bl.append(ptr, 0, 0);
+    EXPECT_EQ(bl.get_num_buffers(), 2);
+    EXPECT_EQ(bl.back().length(), 0);
+    /* rebuild_aligned() call rebuild_aligned_size_and_memory() */
```
Comment (idryomov) [inline src/test/bufferlist.cc:1863]: Nit: calls

**Review by idryomov — COMMENTED** [review]
Looks good to me, just a couple of really minor nits.  I assume you have verified that the new test case still leads to a crash if the patch isn't applied?

**Comment by CongMinYin** [comment]
> Looks good to me, just a couple of really minor nits. I assume you have verified that the new test case still leads to a crash if the patch isn't applied?

Yes, I test the new test case. Without this change in `rebuild_aligned_size_and_memory`, new unit_test failed., after appling this change, new uint_test passed.

had modified comments, please check.

**Review by idryomov — APPROVED** [review]

**Comment by idryomov** [comment]
Merging based on Kefu's earlier runs since the fix itself hasn't changed.
